src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -457,146 +478,174 @@ public Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync()
+                do
+                {
+                    await ComputeAnalyzerDiagnosticsAsync(analysisScope, generateCompilationEvents, getEventQueue, taskToken, cancellationToken).ConfigureAwait(false);
+                } while (_analysisState.HasPendingSymbolAnalysis(analysisScope));
 
-        private async Task ComputeAnalyzerDiagnosticsAsync(AnalysisScope analysisScope, Action generateCompilationEventsOpt, Func<AsyncQueue<CompilationEvent>> getEventQueue, int newTaskToken, CancellationToken cancellationToken)
-        {
-            AnalyzerDriver driver = null;
-            Task computeTask = null;
-            CancellationTokenSource cts;
-            
-            // Generate compilation events, if required.
-            if (generateCompilationEventsOpt != null)
+                // Return computed analyzer diagnostics for the given analysis scope.
+                return _analysisResult.GetDiagnostics(analysisScope, getLocalDiagnostics: true, getNonLocalDiagnostics: false);
+            }
+            catch (Exception e) when (FatalError.ReportUnlessCanceled(e))
             {
-                generateCompilationEventsOpt();
+                throw ExceptionUtilities.Unreachable;
             }
+        }
 
-            // Populate the events cache from the generated compilation events.
-            await PopulateEventsCacheAsync(cancellationToken).ConfigureAwait(false);
-
+        private async Task ComputeAnalyzerDiagnosticsAsync(AnalysisScope analysisScope, Action generateCompilationEventsOpt, Func<AsyncQueue<CompilationEvent>> getEventQueue, int newTaskToken, CancellationToken cancellationToken)
+        {
             try
             {
-                // Get the analyzer driver to execute analysis.
-                driver = await GetAnalyzerDriverAsync(cancellationToken).ConfigureAwait(false);
-
-                // Driver must have been initialized.
-                Debug.Assert(driver.WhenInitializedTask != null);
-                Debug.Assert(!driver.WhenInitializedTask.IsCanceled);
+                AnalyzerDriver driver = null;
+                Task computeTask = null;
+                CancellationTokenSource cts;
+            
+                // Generate compilation events, if required.
+                if (generateCompilationEventsOpt != null)
+                {
+                    generateCompilationEventsOpt();
+                }
 
-                cancellationToken.ThrowIfCancellationRequested();
+                // Populate the events cache from the generated compilation events.
+                await PopulateEventsCacheAsync(cancellationToken).ConfigureAwait(false);
 
-                // Track if this task was suspended by another tree diagnostics request for the same tree.
-                // If so, we wait for the high priority requests to complete before restarting analysis.
-                bool suspendend;
-                do
+                try
```

Comment:
I recommend fusing these try statements. You probably don't want to do the cleanup in the first finally before capturing the failure.